zerver/lib/actions.py

```diff
@@ -4750,9 +4766,13 @@ def gather_subscriptions_helper(user_profile: UserProfile,
 
     recent_traffic = get_streams_traffic(stream_ids=stream_ids)
 
+    # "is_announcement_only" is a valid api field for PATCH /streams/{stream_id},
+    # but not for POST /users/me/subscriptions.
+    subscription_api_fields = Stream.API_FIELDS.copy()
+    subscription_api_fields.remove("is_announcement_only")
     all_streams = get_active_streams(user_profile.realm).select_related(
         "realm").values(
-            *Stream.API_FIELDS,
+            *subscription_api_fields,
```

Comment:
I applied this change as it's a more consistent way to handle this field not present in the database:
```
diff --git a/zerver/lib/actions.py b/zerver/lib/actions.py
index 2fc6f649fc..4dafef5dc2 100644
--- a/zerver/lib/actions.py
+++ b/zerver/lib/actions.py
@@ -3651,6 +3651,10 @@ def do_change_stream_post_policy(stream: Stream, stream_post_policy: int) -> Non
     )
     send_event(stream.realm, event, can_access_stream_user_ids(stream))
 
+    # Backwards-compatibility code: We removed the
+    # is_announcement_only property in early 2020, but we send a
+    # duplicate event for legacy mobile clients that might want the
+    # data.
     event = dict(
         op="update",
         type="stream",
@@ -4766,13 +4770,13 @@ def gather_subscriptions_helper(user_profile: UserProfile,
 
     recent_traffic = get_streams_traffic(stream_ids=stream_ids)
 
-    # "is_announcement_only" is a valid api field for PATCH /streams/{stream_id},
-    # but not for POST /users/me/subscriptions.
-    subscription_api_fields = Stream.API_FIELDS.copy()
-    subscription_api_fields.remove("is_announcement_only")
+    # We removed `is_announcement_only` in early 2020, replacing it
+    # with the more useful stream_post_policy.  As a result,
+    # "is_announcement_only" is a valid api field for PATCH
+    # /streams/{stream_id}, but not for POST /users/me/subscriptions.
     all_streams = get_active_streams(user_profile.realm).select_related(
         "realm").values(
-            *subscription_api_fields,
+            *Stream.API_FIELDS,
             # date_created is used as an input for the is_old_stream computed field.
             "date_created",
             # The realm_id and recipient_id are generally not needed in the API.
@@ -4829,10 +4833,6 @@ def gather_subscriptions_helper(user_profile: UserProfile,
             if field_name == "id":
                 stream_dict['stream_id'] = stream["id"]
                 continue
-            elif field_name == "is_announcement_only":
-                stream_dict['is_announcement_only'] = \
-                    stream['stream_post_policy'] == Stream.STREAM_POST_POLICY_ADMINS
-                continue
             stream_dict[field_name] = stream[field_name]
 
         # Copy Subscription.API_FIELDS except for "active", which is
@@ -4843,6 +4843,11 @@ def gather_subscriptions_helper(user_profile: UserProfile,
         # Backwards-compatibility for clients that haven't been
         # updated for the in_home_view => is_muted API migration.
         stream_dict['in_home_view'] = not stream_dict['is_muted']
+        # Backwards-compatibility for clients that haven't been
+        # updated for the is_announcement_only -> stream_post_policy
+        # migration.
+        stream_dict['is_announcement_only'] = \
+            stream['stream_post_policy'] == Stream.STREAM_POST_POLICY_ADMINS
 
         # Add a few computed fields not directly from the data models.
         stream_dict['is_old_stream'] = is_old_stream(stream["date_created"])
@@ -4886,15 +4891,13 @@ def gather_subscriptions_helper(user_profile: UserProfile,
                 if field_name == "id":
                     stream_dict['stream_id'] = stream["id"]
                     continue
-                elif field_name == "is_announcement_only":
-                    stream_dict['is_announcement_only'] = \
-                        stream['stream_post_policy'] == Stream.STREAM_POST_POLICY_ADMINS
-                    continue
                 stream_dict[field_name] = stream[field_name]
 


             stream_dict['is_old_stream'] = is_old_stream(stream["date_created"])
             stream_dict['stream_weekly_traffic'] = get_average_weekly_stream_traffic(
                 stream["id"], stream["date_created"], recent_traffic)
+            stream_dict['is_announcement_only'] = \
+                stream['stream_post_policy'] == Stream.STREAM_POST_POLICY_ADMINS
 
             if is_public or user_profile.is_realm_admin:
                 subscribers = subscriber_map[stream["id"]]
diff --git a/zerver/models.py b/zerver/models.py
index 9945dfb996..78876990d8 100644
--- a/zerver/models.py
+++ b/zerver/models.py
@@ -1444,7 +1444,6 @@ class Stream(models.Model):
         "rendered_description",
         "invite_only",
         "is_web_public",
-        "is_announcement_only",
         "stream_post_policy",
         "history_public_to_subscribers",
         "first_message_id",
```